### PR TITLE
Update Compression Base Library

### PR DESCRIPTION
**Pull request by panchow**
Improvements include:
- Better Java doc with proper formatting.
- Change parameter names for consistency and easier to understand.
- Made compressionException a checked exception.

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/compression/Compression.java [hunk 1]
```diff
@@ -92,21 +102,23 @@ public interface Compression {
...
-   * @param compressedBuffer The compressed buffer.
+   * @param compressedBuffer The compressed buffer.  It cannot be null or empty.
    * @return Size of the original data.
+   * @throws CompressionException Throws this exception if compression has internal failure.
    */
-  default int getDecompressBufferSize(byte[] compressedBuffer) {
-    if (compressedBuffer == null) {
-      throw new IllegalArgumentException("compressedBuffer cannot be null.");
+  default int getDecompressBufferSize(byte[] compressedBuffer) throws CompressionException {
+    if (compressedBuffer == null || compressedBuffer.length == 0) {
```
Comment (snalli) [inline ambry-api/src/main/java/com/github/ambry/compression/Compression.java:120]: condense into a helper like isCompressedBufferEmpty()

File: ambry-api/src/main/java/com/github/ambry/compression/Compression.java [hunk 2]
```diff
@@ -116,60 +128,56 @@ default int getDecompressBufferSize(byte[] compressedBuffer) {
...
-   * @param sourceData The source uncompressed data to compress.  It cannot be null or empty.
+   * @param sourceBuffer The source uncompressed data to compress.  It cannot be null or empty.
    * @return Pair that contains the compressed buffer and the buffer usage size in bytes.
+   * @throws CompressionException Throws this exception if compression has internal failure.
    */
-  default Pair<Integer, byte[]> compress(byte[] sourceData) {
-    if (sourceData == null || sourceData.length == 0) {
-      throw new IllegalArgumentException("sourceData cannot be null or empty.");
+  default Pair<Integer, byte[]> compress(byte[] sourceBuffer) throws CompressionException {
+    if (sourceBuffer == null || sourceBuffer.length == 0) {
```
Comment (snalli) [inline ambry-api/src/main/java/com/github/ambry/compression/Compression.java:156]: condense into a helper like isSourceBufferEmpty()

File: ambry-api/src/main/java/com/github/ambry/compression/Compression.java [hunk 3]
```diff
@@ -116,60 +128,56 @@ default int getDecompressBufferSize(byte[] compressedBuffer) {
...
+   * @param sourceDataSize Size of source data to read in bytes, not the size of sourceBuffer.
+   * @return Pair that contains the compressed buffer and the buffer usage sourceDataSize in bytes.
+   * @throws CompressionException Throws this exception if compression has internal failure.
    */
-  default Pair<Integer, byte[]> compress(byte[] sourceData, int offset, int size) {
-    if (sourceData == null || sourceData.length == 0) {
-      throw new IllegalArgumentException("sourceData cannot be null or empty.");
+  default Pair<Integer, byte[]> compress(byte[] sourceBuffer, int sourceBufferOffset, int sourceDataSize)
+      throws CompressionException {
+    if (sourceBuffer == null || sourceBuffer.length == 0) {
```
Comment (snalli) [inline ambry-api/src/main/java/com/github/ambry/compression/Compression.java:175]: condense into a helper like isSourceBufferEmpty()

File: ambry-api/src/main/java/com/github/ambry/compression/Compression.java [hunk 4]
```diff
@@ -179,24 +187,26 @@ default Pair<Integer, byte[]> compress(byte[] sourceData, int offset, int size)
...
 
   /**
    * Decompress the compressed buffer.  The compressedBuffer is not shared and contains compressed buffer only.
    * This method allocates and returns the buffer to hold the decompressed data.
    * @param compressedBuffer The compressed buffer generated in compress() method.
    * @return The original/decompressed data.
+   * @throws CompressionException Throws this exception if decompression has internal failure.
    */
-  default byte[] decompress(byte[] compressedBuffer) {
+  default byte[] decompress(byte[] compressedBuffer) throws CompressionException {
```
Comment (snalli) [inline ambry-api/src/main/java/com/github/ambry/compression/Compression.java:209]: condense into a helper like isCompressedBufferEmpty()

File: ambry-api/src/main/java/com/github/ambry/compression/Compression.java [hunk 5]
```diff
@@ -205,21 +215,24 @@ default byte[] decompress(byte[] compressedBuffer) {
...
-   * @param offset Offset in the compressedBuffer to start reading.
-   * @param size Number of bytes to read in compressedBuffer.
+   * @param compressedBufferOffset Offset in the compressedBuffer to start reading.
+   * @param compressedDataSize Number of bytes to read in compressedBuffer, not the size of compressedBuffer.
    * @return The original/decompressed data.
+   * @throws CompressionException Throws this exception if decompression has internal failure.
    */
-  default byte[] decompress(byte[] compressedBuffer, int offset, int size) {
+  default byte[] decompress(byte[] compressedBuffer, int compressedBufferOffset, int compressedDataSize)
+      throws CompressionException {
```
Comment (snalli) [inline ambry-api/src/main/java/com/github/ambry/compression/Compression.java:228]: condense into a helper like isCompressedBufferEmpty()

File: ambry-commons/src/main/java/com/github/ambry/compression/BaseCompression.java [hunk 6]
```diff
@@ -21,9 +21,23 @@
...
+   * The minimal overhead size of the compressed data structure in bytes.
+   * The overhead structure contains:
+   * - 1 (version)
+   * - 1 (name size)
+   * - N (name chars)
+   * - 4 (original size)
+   * Excluding the name chars, the minimal size is 6 bytes.
+   */
+  private static final int SIZE_OF_VERSION_AND_ORIGINAL_SIZE = 5;
+  static final int SIZE_OF_VERSION_AND_ORIGINAL_SIZE_AND_NAME_SIZE = SIZE_OF_VERSION_AND_ORIGINAL_SIZE + 1;
```
Comment (snalli) [inline ambry-commons/src/main/java/com/github/ambry/compression/BaseCompression.java:40]: can u think of a more concise name for this constant ?

File: ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+  /**
+   * Add an instance of compression.  Every compression instance has a unique name.
+   * Decompression will use the compression algorithm name to find the decompression instance.
+   *
+   * @param compressionInstance The compression instance to add.
+   */
+  public void add(Compression compressionInstance) {
+    Objects.requireNonNull(compressionInstance, "compressionInstance");
+
```
Comment (snalli) [inline ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java:51]: rm empty line

File: ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+   * Find the name of the algorithm used to compression the specified compressed buffer.
+   *
+   * @param compressedBuffer The compressed buffer generated by the compress() method.
+   * @param compressedBufferOffset Offset in compressedBuffer to start reading.
+   * @param compressedDataSize The size of compressed data in compressedBuffer.
+   * @return Name of the compression algorithm.
+   */
+  public String getAlgorithmName(byte[] compressedBuffer, int compressedBufferOffset, int compressedDataSize) {
+    BaseCompression.verifyCompressedBuffer(compressedBuffer, compressedBufferOffset, compressedDataSize);
+
```
Comment (snalli) [inline ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java:80]: rm empty line

File: ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+   * Find the decompression instance based on the algorithm name stored in compressedBuffer.
+   * The entire buffer is dedicated to the compressed data.
+   *
+   *
+   * @param compressedBuffer The compressed buffer generated by the compress() method.
+   * @return The decompressor that can decompress this compressed buffer.  Null if no decompressor found.
+   */
+  public Compression getByCompressedData(byte[] compressedBuffer) {
+    Objects.requireNonNull(compressedBuffer, "compressedBuffer");
+
```
Comment (snalli) [inline ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java:96]: rm empty line

**Review by snalli — CHANGES_REQUESTED** [review]
requesting minor changes

File: ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java (hunk 9 above)
Comment (panchow) [reply ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java:96]: Regarding the empty lines after verification.  This is a nit and depends on personal style preference.  Some people, including myself, like to group code by role.  I prefer to group all parameter checks near the top of the function, followed by an empty line, and then the core code.  Unless there is a specific reason while an empty line is not recommended between parameter checks and core code, I would like to keep the empty line there.

File: ambry-api/src/main/java/com/github/ambry/compression/Compression.java (hunk 1 above)
Comment (panchow) [reply ambry-api/src/main/java/com/github/ambry/compression/Compression.java:120]: Even better, I'd add a method to Utils just to verify empty byte array.  Will appear in next update.

File: ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java (hunk 9 above)
Comment (snalli) [reply ambry-commons/src/main/java/com/github/ambry/compression/CompressionMap.java:96]: It's not clear however, what the purpose of empty line  itself is, which is to separate logically un-related code known only to the author. A comment, preceded and followed by an empty line, would be a better separator.

**Review by snalli — APPROVED** [review]

**Comment by snalli** [comment]
approved, will merge once tests pass

**Pull request merged**